src/Compilers/Core/MSBuildTask/MvidReader.cs

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+
+namespace Microsoft.CodeAnalysis.BuildTasks
+{
+    public sealed class MvidReader : BinaryReader
+    {
+        public MvidReader(Stream stream) : base(stream)
+        {
+        }
+
+        public static Guid ReadAssemblyMvid(Stream stream)
+        {
+            try
+            {
+                var mvidReader = new MvidReader(stream);
+                return mvidReader.FindMvid();
+            }
+            catch (EndOfStreamException)
+            {
+            }
+
+            return Guid.Empty;
+        }
+
+        public Guid FindMvid()
+        {
+            Guid empty = Guid.Empty;
+
+            // DOS Header (64), DOS Stub (64), PE Signature (4), COFF Header (20), PE Header (224)
+            if (BaseStream.Length < 64 + 64 + 4 + 20 + 224)
+            {
+                return empty;
+            }
+
+            // DOS Header: PE (2)
+            if (ReadUInt16() != 0x5a4d)
+            {
+                return empty;
+            }
+
+            // DOS Header: Start (58)
+            Skip(58);
+
+            // DOS Header: Address of PE Signature
+            MoveTo(ReadUInt32());
+
+            // PE Signature ('P' 'E' null null)
+            if (ReadUInt32() != 0x00004550)
+            {
+                return empty;
+            }
+
+            // COFF Header: Machine (2)
+            Skip(2);
+
+            // COFF Header: NumberOfSections (2)
+            ushort sections = ReadUInt16();
+
+            // COFF Header: TimeDateStamp (4), PointerToSymbolTable (4), NumberOfSymbols (4)
+            Skip(12);
+
+            // COFF Header: OptionalHeaderSize (2)
+            ushort optionalHeaderSize = ReadUInt16();
+
+            // COFF Header: Characteristics (2)
+            Skip(2);
+
+            // Optional header
+            Skip(optionalHeaderSize);
+
+            // Section headers
+            return FindMvidInSections(sections);
+        }
+
+        string ReadSectionName()
+        {
+            int length = 8;
+            int read = 0;
+            var buffer = new char[length];
+            var bytes = ReadBytes(length);
+            while (read < length)
+            {
+                var current = bytes[read];
+                if (current == 0)
+                    break;
+
+                buffer[read++] = (char)current;
+            }
+
+            return new string(buffer, 0, read);
+        }
+
+        Guid FindMvidInSections(ushort count)
+        {
+            for (int i = 0; i < count; i++)
+            {
+                // Section: Name (8)
+                string name = ReadSectionName();
```

Comment:
>string name = ReadSectionName(); [](start = 16, length = 32)

Seems like this could be simplified quite a bit. 

Something like 

bytes[] name = ReadBytes(8);
if (name.Length == 8 && name[0] = '.' && name[1] == 'm' && name[2] == 'v' ...) 
{
   read data
} 
else
{
   Skip(size_of_section_header - 8);
}